src/recommendation/src/test/scala/SARSpec.scala

```diff
@@ -0,0 +1,302 @@
+        val groundTrueScore = row.getAs[String](itemJ).toFloat
+        val sparkSarScore = simMap.getOrElse(itemI, Map()).getOrElse(itemJ, "-1")
+        assert(groundTrueScore == sparkSarScore)
+      })
+    })
+    (model, userMapBC, itemMapBC)
+  }
+
+  def test_product_recommendations(tlcSampleData: DataFrame, threshold: Int, similarityFunction: String,
+    simFile: String, user_aff: String,
+    userPredFile: String): Unit = {
+    val session = tlcSampleData.sparkSession
+    val sparkContext = session.sparkContext
+
+    val (model, userMapBC, itemMapBC) = test_affinity_matrices(tlcSampleData, threshold, similarityFunction, simFile,
+      user_aff)
+
+    val recoverUser = udf((userID: Integer) => userMapBC.value.getOrElse[String](userID, "-1"))
+    val recoverItem = udf((itemID: Integer) => itemMapBC.value.getOrElse[String](itemID, "-1"))
+
+    val usersProducts = tlcSampleData
+      .filter(col("userId") === "0003000098E85347")
+      .select("productId")
+      .distinct()
+      .collect()
+      .map(_.getString(0))
+
+    val usersProductsBC = sparkContext.broadcast(usersProducts)
+
+    val users = model.recommendForAllUsers(5 + usersProducts.length)
+
+    val filterScore = udf((items: Seq[Int], ratings: Seq[Float]) => {
+      items.zipWithIndex
+        .filter(p => {
+          val itemId = itemMapBC.value.getOrElse[String](p._1, "-1")
+          val bol = usersProductsBC.value.contains(itemId)
+          !bol
+        }).map(p => (p._1, ratings.toList(p._2)))
+    })
+
+    val temp = users
+      .select(col("customerID"), filterScore(col("recommendations.itemID"), col("recommendations.rating")) as
+        "recommendations")
+      .select(col("customerID"), col("recommendations._1") as "itemID", col("recommendations._2") as "rating")
+      .select(
+        recoverUser(col("customerID")) as "customerID",
+        recoverItem(col("itemID")(0)) as "rec1",
+        recoverItem(col("itemID")(1)) as "rec2",
+        recoverItem(col("itemID")(2)) as "rec3",
+        recoverItem(col("itemID")(3)) as "rec4",
+        recoverItem(col("itemID")(4)) as "rec5",
+        col("rating")(0) as "score1",
+        col("rating")(1) as "score2",
+        col("rating")(2) as "score3",
+        col("rating")(3) as "score4",
+        col("rating")(4) as "score5")
+
+    val rows = temp
+      .filter(col("customerID") === "0003000098E85347")
+      .take(1)
+
+    val answer = session.read.option("header", "true").csv(userPredFile).collect()
+
+    assert(rows(0).getString(0) == "0003000098E85347", "Assert Customer ID's Match")
+
+    assert(rows(0).getString(1) == answer(0).getString(1))
+    assert(rows(0).getString(2) == answer(0).getString(2))
+    assert(rows(0).getString(3) == answer(0).getString(3))
+    assert(rows(0).getString(4) == answer(0).getString(4))
+    assert(rows(0).getString(5) == answer(0).getString(5))
+    assert("%.3f".format(rows(0).getFloat(6)) == "%.3f".format(answer(0).getString(11).toFloat))
```

Comment:
if you are trying to do tolerant comparisions, take a look at the DataFrameEquality trait. It should have  row equality and a tolerant double equality thats very customizable. You hook into it with === instead of ==